File: src/heapsort.py
def heapify(heap, index, size):
    greatest = index
    left = (2 * index) + 1
    right = (2 * index) + 2
    
    if left < size and heap[left] > heap[greatest]:
        greatest = left
    if right < size and heap[right] > heap[greatest]:
        greatest = right
    if greatest != index:
        heap[greatest], heap[index] = heap[index], heap[greatest]
        heapify(heap, greatest, size)

def heapsort(heap):
    size = len(heap)
    
    for x in range((size // 2) - 1, -1, -1):
        heapify(heap, x, size)
    
    for x in range(size - 1, 0, -1):
        heap[0], heap[x] = heap[x], heap[0]
        heapify(heap, 0, x)

if __name__ == '__main__':
    arr = [1212, 123, 1756, 12732, 849, 192, 3, 7592, 8503]
    heapsort(arr)
    
    print(arr)

File: static/sorting/heapsort.js
// This is the code for the heap sort, it builds a heap from the array and then repeatedly extracts the top element from the heap
function heapify(nums, index, size) {
    greatest = index;
    left = 2 * index + 1;
    right = 2 * index + 2;

    if (left < size && nums[left] > nums[greatest]) {
        greatest = left;
    }
    if (right < size && nums[right] > nums[greatest]) {
        greatest = right;
    }
    if (greatest != index) {
        [nums[index], nums[greatest]] = [nums[greatest], nums[index]];
        heapify(nums, greatest, size);
    }
}

function heapsort(nums) {
    size = nums.length;

    for (let x = Math.floor(size / 2) - 1; x >= 0; x--) {
        heapify(nums, x, size);
    }

    for (let x = size - 1; x > 0; x--) {
        [nums[0], nums[x]] = [nums[x], nums[0]];
        heapify(nums, 0, x);
    }
}
